### PR TITLE
Fixes #2255: leaked transactions in ConsistentKeyLocker

### LINKED ISSUE #2255: Leaked transactions in ConsistentKeyLocker
Transactions created in ConsistentKeyLocker.overrideTimestamp() are neither committed or rolled back by the callers. This results in leaked transactions. ConsistentKeyLocker needs to be updated to commit or rollback the created transactions.
 
Relevant code from ConsistentKeyLocker.java given below.

https://github.com/JanusGraph/janusgraph/blob/9ba0b2d850066a32551f9da4af0780f643288376/janusgraph-core/src/main/java/org/janusgraph/diskstorage/locking/consistentkey/ConsistentKeyLocker.java#L549-L554

https://github.com/JanusGraph/janusgraph/blob/9ba0b2d850066a32551f9da4af0780f643288376/janusgraph-core/src/main/java/org/janusgraph/diskstorage/locking/consistentkey/ConsistentKeyLocker.java#L384-L388

https://github.com/JanusGraph/janusgraph/blob/9ba0b2d850066a32551f9da4af0780f643288376/janusgraph-core/src/main/java/org/janusgraph/diskstorage/locking/consistentkey/ConsistentKeyLocker.java#L400-L403

https://github.com/JanusGraph/janusgraph/blob/9ba0b2d850066a32551f9da4af0780f643288376/janusgraph-core/src/main/java/org/janusgraph/diskstorage/locking/consistentkey/ConsistentKeyLocker.java#L535-L539

### PR DESCRIPTION
**Pull request by mneethiraj**
#2225: Fixed transaction leaks in ConsistentKeyLocker. Updated unit tests to detect leaked transaction.

(cherry picked from commit 12718473fae6e9172e292cfcd61de175d22b2577)
(cherry picked from commit 3e16a8fdd8dc81a47529cf0950d60c7ed3debdd3)

### CODE REVIEW

**Review by li-boxuan — APPROVED** [review]

**Comment by mneethiraj** [comment]
Boxuan,

 

Thanks for reviewing and approving the pull request. Is there anything else needed from me before this can be merged?

 

Thanks,

Madhan

**Comment by li-boxuan** [comment]
Hi Madhan, there is nothing pending from your side. According to https://docs.janusgraph.org/development/#review-then-commit-rtc, we need one more approval from another committer. If there is no review in the next two days, I will merge your PR following the lazy consensus.

File: janusgraph-core/src/main/java/org/janusgraph/diskstorage/locking/consistentkey/ConsistentKeyLocker.java
Comment (porunov) [inline janusgraph-core/src/main/java/org/janusgraph/diskstorage/locking/consistentkey/ConsistentKeyLocker.java:401]: Why do we adding `new Throwable()` to the log? I guess, it isn't useful because it doesn't have any information. Instead, I would log `toString()` of the transaction, which may have more useful information. I.e. something like:
```
if (log.isDebugEnabled()) {
    log.debug("Transaction is still open. Rolling back the transaction: "+tx.toString());
}
```

File: janusgraph-core/src/main/java/org/janusgraph/diskstorage/locking/consistentkey/ConsistentKeyLocker.java
Comment (porunov) [inline janusgraph-core/src/main/java/org/janusgraph/diskstorage/locking/consistentkey/ConsistentKeyLocker.java:406]: In this situation the transaction will leak. I think we should log here `error` with the explanation that the above transaction is leaked. Of course it is better to somehow track it and let the logic somehow handle the transaction but I guess it makes the implementation slightly harder, so I'm totally fine to just log an error here and add a `TODO` message to actually resolve it later.
Something like:
```
//TODO: We need to handle the transaction newTx which potentially may leak into memory
log.error("Failed to rollback the transaction "+newTx.toString()+". The transaction now may be leaked into memory.", e);
```

File: janusgraph-core/src/main/java/org/janusgraph/diskstorage/locking/consistentkey/ConsistentKeyLocker.java
Comment (porunov) [inline janusgraph-core/src/main/java/org/janusgraph/diskstorage/locking/consistentkey/ConsistentKeyLocker.java:439]: We have multiple same `finally` blocks. I propose to move the same code into a separate method. I.e.:
```
private void rollbackIfNotNull(BaseTransaction tx){
    if (tx != null) {
        try {
            if (log.isDebugEnabled()) {
                log.debug("Transaction is still open. Rolling back the transaction: "+tx.toString());
            }
            tx.rollback();
        } catch (Throwable e) {
            //TODO: We need to handle the transaction newTx which potentially may leak into memory
            log.error("Failed to rollback the transaction "+tx.toString()+". The transaction now may be leaked into memory.", e);
        }
    }
}
```

After that we can change all places where you added `finally` like the next:
```
finally {
  rollbackIfNotNull(newTx);
}
```

**Review by porunov — CHANGES_REQUESTED** [review]
Thank you @mneethiraj for the contribution! I have several comments below

File: janusgraph-core/src/main/java/org/janusgraph/diskstorage/locking/consistentkey/ConsistentKeyLocker.java
Comment (mneethiraj) [reply janusgraph-core/src/main/java/org/janusgraph/diskstorage/locking/consistentkey/ConsistentKeyLocker.java:401]: Throwable in this debug log statement will print the stack, which could be useful in troubleshooting leaked transactions. Let me know if this should be removed.

I agree on including tx.toString() in the log. Will update shortly for this.

File: janusgraph-core/src/main/java/org/janusgraph/diskstorage/locking/consistentkey/ConsistentKeyLocker.java
Comment (mneethiraj) [reply janusgraph-core/src/main/java/org/janusgraph/diskstorage/locking/consistentkey/ConsistentKeyLocker.java:406]: I agree. Will update the patch shortly. Thanks!

File: janusgraph-core/src/main/java/org/janusgraph/diskstorage/locking/consistentkey/ConsistentKeyLocker.java
Comment (mneethiraj) [reply janusgraph-core/src/main/java/org/janusgraph/diskstorage/locking/consistentkey/ConsistentKeyLocker.java:439]: I agree. Will update the patch shortly. Thanks!

File: janusgraph-core/src/main/java/org/janusgraph/diskstorage/locking/consistentkey/ConsistentKeyLocker.java
Comment (porunov) [reply janusgraph-core/src/main/java/org/janusgraph/diskstorage/locking/consistentkey/ConsistentKeyLocker.java]: Good point. Agree about stack trace. It might be useful

**Review by porunov — COMMENTED** [review]
The code LGTM but could you please squash you commits into one commit?
You can use `git rebase -i a22a62e01271d1e3bdeeb6f698a3a3fed641a7c1` where you will need to pick one commit and squash another commit or you can use any other technique to squash commits which you are convenient with

**Review by porunov — APPROVED** [review]
LGTM. Thank you @mneethiraj for the contribution!

**Pull request merged**
